File: clients/typescript/src/apis/DuplicatesApi.ts
/* tslint:disable */
/* eslint-disable */

import * as runtime from '../runtime';
import {
    type DuplicatesRequest,
    DuplicatesRequestFromJSON,
    DuplicatesRequestToJSON,
} from '../models/DuplicatesRequest';
import {
    type DuplicatesResponse,
    DuplicatesResponseFromJSON,
    DuplicatesResponseToJSON,
} from '../models/DuplicatesResponse';
import {
    type HTTPValidationError,
    HTTPValidationErrorFromJSON,
    HTTPValidationErrorToJSON,
} from '../models/HTTPValidationError';

export interface FindDuplicatesRequest {

    duplicatesRequest: DuplicatesRequest;
}

export interface SuggestedDedupThresholdRequest {

    customerId?: string | null;
}

export class DuplicatesApi extends runtime.BaseAPI {

    async findDuplicatesRequestOpts(requestParameters: FindDuplicatesRequest): Promise<runtime.RequestOpts> {
        if (requestParameters['duplicatesRequest'] == null) {
            throw new runtime.RequiredError(
                'duplicatesRequest',
                'Required parameter "duplicatesRequest" was null or undefined when calling findDuplicates().'
            );
        }

        const queryParameters: any = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("bearerAuth", []);

            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }

        let urlPath = `/v1/duplicates`;

        return {
            path: urlPath,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: DuplicatesRequestToJSON(requestParameters['duplicatesRequest']),
        };
    }

    async findDuplicatesRaw(requestParameters: FindDuplicatesRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<DuplicatesResponse>> {
        const requestOptions = await this.findDuplicatesRequestOpts(requestParameters);
        const response = await this.request(requestOptions, initOverrides);

        return new runtime.JSONApiResponse(response, (jsonValue) => DuplicatesResponseFromJSON(jsonValue));
    }

    async findDuplicates(requestParameters: FindDuplicatesRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<DuplicatesResponse> {
        const response = await this.findDuplicatesRaw(requestParameters, initOverrides);
        return await response.value();
    }

    async suggestedDedupThresholdRequestOpts(requestParameters: SuggestedDedupThresholdRequest): Promise<runtime.RequestOpts> {
        const queryParameters: any = {};

        if (requestParameters['customerId'] != null) {
            queryParameters['customer_id'] = requestParameters['customerId'];
        }

        const headerParameters: runtime.HTTPHeaders = {};

        if (this.configuration && this.configuration.accessToken) {
            const token = this.configuration.accessToken;
            const tokenString = await token("bearerAuth", []);

            if (tokenString) {
                headerParameters["Authorization"] = `Bearer ${tokenString}`;
            }
        }

        let urlPath = `/v1/duplicates/suggested-threshold`;

        return {
            path: urlPath,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        };
    }

    async suggestedDedupThresholdRaw(requestParameters: SuggestedDedupThresholdRequest, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<runtime.ApiResponse<any>> {
        const requestOptions = await this.suggestedDedupThresholdRequestOpts(requestParameters);
        const response = await this.request(requestOptions, initOverrides);

        if (this.isJsonMime(response.headers.get('content-type'))) {
            return new runtime.JSONApiResponse<any>(response);
        } else {
            return new runtime.TextApiResponse(response) as any;
        }
    }

    async suggestedDedupThreshold(requestParameters: SuggestedDedupThresholdRequest = {}, initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<any> {
        const response = await this.suggestedDedupThresholdRaw(requestParameters, initOverrides);
        return await response.value();
    }

}

File: clients/rust/src/apis/duplicates_api.rs
use reqwest;
use serde::{Deserialize, Serialize, de::Error as _};
use crate::{apis::ResponseContent, models};
use super::{Error, configuration, ContentType};


#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum FindDuplicatesError {
    Status422(models::HttpValidationError),
    UnknownValue(serde_json::Value),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum SuggestedDedupThresholdError {
    Status422(models::HttpValidationError),
    UnknownValue(serde_json::Value),
}


pub async fn find_duplicates(configuration: &configuration::Configuration, duplicates_request: models::DuplicatesRequest) -> Result<models::DuplicatesResponse, Error<FindDuplicatesError>> {
    let p_body_duplicates_request = duplicates_request;

    let uri_str = format!("{}/v1/duplicates", configuration.base_path);
    let mut req_builder = configuration.client.request(reqwest::Method::POST, &uri_str);

    if let Some(ref user_agent) = configuration.user_agent {
        req_builder = req_builder.header(reqwest::header::USER_AGENT, user_agent.clone());
    }
    if let Some(ref token) = configuration.bearer_access_token {
        req_builder = req_builder.bearer_auth(token.to_owned());
    };
    req_builder = req_builder.json(&p_body_duplicates_request);

    let req = req_builder.build()?;
    let resp = configuration.client.execute(req).await?;

    let status = resp.status();
    let content_type = resp
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("application/octet-stream");
    let content_type = super::ContentType::from(content_type);

    if !status.is_client_error() && !status.is_server_error() {
        let content = resp.text().await?;
        match content_type {
            ContentType::Json => serde_json::from_str(&content).map_err(Error::from),
            ContentType::Text => return Err(Error::from(serde_json::Error::custom("Received `text/plain` content type response that cannot be converted to `models::DuplicatesResponse`"))),
            ContentType::Unsupported(unknown_type) => return Err(Error::from(serde_json::Error::custom(format!("Received `{unknown_type}` content type response that cannot be converted to `models::DuplicatesResponse`")))),
        }
    } else {
        let content = resp.text().await?;
        let entity: Option<FindDuplicatesError> = serde_json::from_str(&content).ok();
        Err(Error::ResponseError(ResponseContent { status, content, entity }))
    }
}

pub async fn suggested_dedup_threshold(configuration: &configuration::Configuration, customer_id: Option<&str>) -> Result<serde_json::Value, Error<SuggestedDedupThresholdError>> {
    let p_query_customer_id = customer_id;

    let uri_str = format!("{}/v1/duplicates/suggested-threshold", configuration.base_path);
    let mut req_builder = configuration.client.request(reqwest::Method::GET, &uri_str);

    if let Some(ref param_value) = p_query_customer_id {
        req_builder = req_builder.query(&[("customer_id", &param_value.to_string())]);
    }
    if let Some(ref user_agent) = configuration.user_agent {
        req_builder = req_builder.header(reqwest::header::USER_AGENT, user_agent.clone());
    }
    if let Some(ref token) = configuration.bearer_access_token {
        req_builder = req_builder.bearer_auth(token.to_owned());
    };

    let req = req_builder.build()?;
    let resp = configuration.client.execute(req).await?;

    let status = resp.status();
    let content_type = resp
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("application/octet-stream");
    let content_type = super::ContentType::from(content_type);

    if !status.is_client_error() && !status.is_server_error() {
        let content = resp.text().await?;
        match content_type {
            ContentType::Json => serde_json::from_str(&content).map_err(Error::from),
            ContentType::Text => return Err(Error::from(serde_json::Error::custom("Received `text/plain` content type response that cannot be converted to `serde_json::Value`"))),
            ContentType::Unsupported(unknown_type) => return Err(Error::from(serde_json::Error::custom(format!("Received `{unknown_type}` content type response that cannot be converted to `serde_json::Value`")))),
        }
    } else {
        let content = resp.text().await?;
        let entity: Option<SuggestedDedupThresholdError> = serde_json::from_str(&content).ok();
        Err(Error::ResponseError(ResponseContent { status, content, entity }))
    }
}

